src/backend/dx12/src/command.rs

```diff
@@ -630,23 +861,55 @@ impl com::RawCommandBuffer<Backend> for CommandBuffer {
             }
         }
         for (set, table) in sets.iter().zip(layout.tables[first_set..].iter()) {
-            set.first_gpu_view.map(|gpu| unsafe {
+            set.first_gpu_view.map(|gpu| {
                 assert!(table.contains(n::SRV_CBV_UAV));
-                self.raw.SetGraphicsRootDescriptorTable(table_id, gpu);
+
+                let root_offset = table_id; // TODO: take push constants into account
+                // Cast is safe as offset **must** be in u32 range. Unable to
+                // create heaps with more descriptors.
+                let table_offset = (gpu.ptr - srv_cbv_uav_base.ptr) as u32;
+                self
+                    .gr_pipeline
+                    .user_data
+                    .set_srv_cbv_uav_table(root_offset, table_offset);
+
                 table_id += 1;
             });
-            set.first_gpu_sampler.map(|gpu| unsafe {
+            set.first_gpu_sampler.map(|gpu| {
                 assert!(table.contains(n::SAMPLERS));
-                self.raw.SetGraphicsRootDescriptorTable(table_id, gpu);
+
+                let root_offset = table_id; // TODO: take push constants into account
+                // Cast is safe as offset **must** be in u32 range. Unable to
+                // create heaps with more descriptors.
+                let table_offset = (gpu.ptr - sampler_base.ptr) as u32;
+                self
+                    .gr_pipeline
+                    .user_data
+                    .set_sampler_table(root_offset, table_offset);
+
                 table_id += 1;
             });
         }
     }
 
     fn bind_compute_pipeline(&mut self, pipeline: &n::ComputePipeline) {
         unsafe {
+            match self.comp_pipeline.pipeline {
+                Some((_, signature)) if signature == pipeline.signature => {
```

Comment:
just to clarify: so if the current pipeline is graphics, but the last compute pipeline matches this one, we still don't need to rebind anything?

Reply:
Pipelines only get bound on draw/dispatch calls to avoid redundant rebinds. We only need/know which pipeline to bind at these calls.
e.g
```
BindGraphicsPipeline() // the pipeline bind here would be useless
... // Fill graphics user data
BindComputePipeline()
... // Fill compute user data
dispatch()
dispatch()
draw()
```

Reply:
I will add a comment as it's sensible question

Reply:
Ok, now I think I understood the question correctly. As far as I can see from the dx12 docs, the graphics and compute root signature are independent so they shouldn't affect each other. The current pipeline state is independent from the root signature stuff afaik.

> If a root signature is changed on a command list, all previous root signature bindings become stale and all newly expected bindings must be set before Draw/Dispatch